Pure Aggregate.ref/one/many require RefEntity with runtime `get`, so pure Entity.make definitions can't build edges

## Summary

In `@effect-dynamodb/schema`, the pure aggregate edge constructors (`Aggregate.ref` / `Aggregate.one` / `Aggregate.many`) type their entity argument as `RefEntity`, whose interface **requires a runtime `get` method**. But the pure `Entity.make` (same package) returns an `EntityDefinition` that does **not** expose `get` (it's a runtime-only operation). As a result you cannot build aggregate edges from pure entity definitions at the type level — which is exactly the AWS-free authoring path the schema/runtime split (#62) exists to enable.

## Reproduction

All within `@effect-dynamodb/schema` (no runtime package):

```ts
import { Aggregate, DynamoModel, Entity } from '@effect-dynamodb/schema'
import { Schema } from 'effect'

class Team extends Schema.Class<Team>('Team')({
  id: Schema.String.pipe(Schema.brand('TeamId')),
  name: Schema.String
}) {}

const Teams = Entity.make({
  model: DynamoModel.configure(Team, { id: { field: 'teamId', identifier: true } }),
  entityType: 'Team',
  primaryKey: { pk: { field: 'pk', composite: ['id'] }, sk: { field: 'sk', composite: [] } }
})

Aggregate.ref(Teams) // ❌ type error
```

```
error TS2345: Argument of type 'EntityDefinition<...>' is not assignable to
parameter of type 'RefEntity<Top, unknown>'.
  Property 'get' is missing in type 'EntityDefinition<...>' but required in
  type 'RefEntity<Top, unknown>'.
```

Same failure for `Aggregate.one('x', { entity: Teams })` and `Aggregate.many('x', { entity: Teams })`.

## Root cause

`internal/AggregateEdges.d.ts` — `RefEntity` requires `get`:

```ts
export interface RefEntity<M extends Schema.Top = Schema.Top, TIndexes = unknown> {
  readonly _tag: "Entity"
  readonly entityType: string
  readonly model: M
  readonly indexes: TIndexes
  readonly schemas: { readonly recordSchema: Schema.Codec<any> }
  /** Get an entity by key — uses `any` for structural compatibility with Entity's narrower key type */
  readonly get: (key: any) => any   // <-- runtime-only op required by the *pure* edge constructors
}
```

The edge derivation only reads `model` / `indexes` / `entityType` at runtime to recover the ref-id field and its (branded) identifier type — it never calls `get`. So `get` is over-required: it forces a runtime `BoundEntity` where a pure `EntityDefinition` should suffice.

## Impact

The pure path (define entities with `@effect-dynamodb/schema` `Entity.make`, build the aggregate edge graph, derive `inputSchema`/`createSchema`/`updateSchema` for an HTTP contract or codegen — all AWS-free) does not typecheck. Consumers must either cast entities to `RefEntity` (erasing the branded-id inference, though #61 already collapses aggregate ref-ids to `string`) or pull the runtime package (re-introducing the AWS SDK the split was meant to keep out).

## Suggested fix

Drop `get` from the `RefEntity` structural bound (and from `OneEdge`/`ManyEdge`/`RefEdge` entity constraints), or make it optional, so a pure `EntityDefinition` satisfies it. The runtime layer can keep its own narrower entity type with `get` for operations; the pure edge constructors only need `{ _tag, entityType, model, indexes, schemas }`.

## Related

- #62 (schema/runtime split — this blocks its intended pure-authoring use case)
- #61 (aggregate ref-id de-branding — same edge-derivation area)

## Versions

- @effect-dynamodb/schema 1.9.1
- effect 4.0.0-beta.85
- typescript 5.9.x, strict

--- a/packages/effect-dynamodb-geo/CHANGELOG.md
+++ b/packages/effect-dynamodb-geo/CHANGELOG.md
@@ -1,5 +1,29 @@
 # @effect-dynamodb/geo
 
+## 1.9.2
+
+### Patch Changes
+
+- fix(schema): make the AWS-free pure-authoring path actually usable (closes #66, closes #67).
+
+  Two follow-ups to the #62 schema/runtime split, both blocking its headline use case
+  (deriving a typed aggregate input/create payload from `@effect-dynamodb/schema` with
+  no AWS SDK):
+  - **#66** — the pure edge constructors (`Aggregate.ref` / `one` / `many`) required a
+    `RefEntity` with a runtime `get` method, so aggregate edges could not be authored
+    from pure `Entity.make` definitions (which have no `get`). `RefEntity` is now the
+    minimal structural bound used only for derivation (`_tag`/`entityType`/`model`/
+    `indexes`/`schemas`); the runtime ref-hydration narrows back to a `get`-bearing
+    entity at its single call site.
+  - **#67** — `deriveAggregateSchemas` (the table-free derivation entry point) returned
+    `Schema.Top` members, so `typeof result.inputSchema.Type` collapsed to `unknown`.
+    It is now generic and returns `Schema.Codec<AggregateInputType<…>>` (plus a
+    `createSchema` alias), so the table-free path is as typed as the top-level
+    `Aggregate.make` — no stub `table` tag or GSI key config needed.
+
+  Type-checked regression tests for both land in the schema package's `tsconfig.test.json`
+  gate (now wired into `pnpm check`).
+
 ## 1.9.1
 
 ### Patch Changes

--- a/packages/effect-dynamodb-geo/package.json
+++ b/packages/effect-dynamodb-geo/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@effect-dynamodb/geo",
-  "version": "1.9.1",
+  "version": "1.9.2",
   "description": "Geospatial index and search for effect-dynamodb using H3 hexagonal grid",
   "type": "module",
   "license": "MIT",

--- a/packages/effect-dynamodb/CHANGELOG.md
+++ b/packages/effect-dynamodb/CHANGELOG.md
@@ -1,5 +1,32 @@
 # effect-dynamodb
 
+## 1.9.2
+
+### Patch Changes
+
+- fix(schema): make the AWS-free pure-authoring path actually usable (closes #66, closes #67).
+
+  Two follow-ups to the #62 schema/runtime split, both blocking its headline use case
+  (deriving a typed aggregate input/create payload from `@effect-dynamodb/schema` with
+  no AWS SDK):
+  - **#66** — the pure edge constructors (`Aggregate.ref` / `one` / `many`) required a
+    `RefEntity` with a runtime `get` method, so aggregate edges could not be authored
+    from pure `Entity.make` definitions (which have no `get`). `RefEntity` is now the
+    minimal structural bound used only for derivation (`_tag`/`entityType`/`model`/
+    `indexes`/`schemas`); the runtime ref-hydration narrows back to a `get`-bearing
+    entity at its single call site.
+  - **#67** — `deriveAggregateSchemas` (the table-free derivation entry point) returned
+    `Schema.Top` members, so `typeof result.inputSchema.Type` collapsed to `unknown`.
+    It is now generic and returns `Schema.Codec<AggregateInputType<…>>` (plus a
+    `createSchema` alias), so the table-free path is as typed as the top-level
+    `Aggregate.make` — no stub `table` tag or GSI key config needed.
+
+  Type-checked regression tests for both land in the schema package's `tsconfig.test.json`
+  gate (now wired into `pnpm check`).
+
+- Updated dependencies
+  - @effect-dynamodb/schema@1.9.2
+
 ## 1.9.1
 
 ### Patch Changes

--- a/packages/effect-dynamodb/package.json
+++ b/packages/effect-dynamodb/package.json
@@ -1,6 +1,6 @@
 {
   "name": "effect-dynamodb",
-  "version": "1.9.1",
+  "version": "1.9.2",
   "description": "Effect TS ORM for DynamoDB — schema-driven entity modeling, single-table design, composite keys, transactions, and Stream-based pagination",
   "type": "module",
   "license": "MIT",

--- a/packages/effect-dynamodb/src/Aggregate.ts
+++ b/packages/effect-dynamodb/src/Aggregate.ts
@@ -1391,9 +1391,15 @@ const hydrateAggregateRefs = (
       // Deduplicate IDs for this entity type
       const uniqueIds = [...new Set(reqs.map((r) => r.id))]
 
+      // `RefEntity` is the pure structural bound (no `get` — issue #66) so that
+      // aggregate edges can be authored from AWS-free entity definitions. Ref
+      // hydration is a runtime operation, so narrow back to the get-bearing
+      // runtime entity here — the runtime path always supplies a real entity.
+      const runtimeEntity = entity as RefEntity & { readonly get: (key: any) => any }
+
       // Build EntityGet intermediates for batch fetching
       const getOps = uniqueIds.map(
-        (id) => entity.get({ [identifierName]: id }) as EntityGet<any, any, any, any>,
+        (id) => runtimeEntity.get({ [identifierName]: id }) as EntityGet<any, any, any, any>,
       )
 
       const results = yield* Batch.get(getOps)

--- a/packages/language-service/CHANGELOG.md
+++ b/packages/language-service/CHANGELOG.md
@@ -1,5 +1,29 @@
 # @effect-dynamodb/language-service
 
+## 1.9.2
+
+### Patch Changes
+
+- fix(schema): make the AWS-free pure-authoring path actually usable (closes #66, closes #67).
+
+  Two follow-ups to the #62 schema/runtime split, both blocking its headline use case
+  (deriving a typed aggregate input/create payload from `@effect-dynamodb/schema` with
+  no AWS SDK):
+  - **#66** — the pure edge constructors (`Aggregate.ref` / `one` / `many`) required a
+    `RefEntity` with a runtime `get` method, so aggregate edges could not be authored
+    from pure `Entity.make` definitions (which have no `get`). `RefEntity` is now the
+    minimal structural bound used only for derivation (`_tag`/`entityType`/`model`/
+    `indexes`/`schemas`); the runtime ref-hydration narrows back to a `get`-bearing
+    entity at its single call site.
+  - **#67** — `deriveAggregateSchemas` (the table-free derivation entry point) returned
+    `Schema.Top` members, so `typeof result.inputSchema.Type` collapsed to `unknown`.
+    It is now generic and returns `Schema.Codec<AggregateInputType<…>>` (plus a
+    `createSchema` alias), so the table-free path is as typed as the top-level
+    `Aggregate.make` — no stub `table` tag or GSI key config needed.
+
+  Type-checked regression tests for both land in the schema package's `tsconfig.test.json`
+  gate (now wired into `pnpm check`).
+
 ## 1.9.1
 
 ### Patch Changes

--- a/packages/language-service/package.json
+++ b/packages/language-service/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@effect-dynamodb/language-service",
-  "version": "1.9.1",
+  "version": "1.9.2",
   "description": "TypeScript Language Service Plugin for effect-dynamodb — shows DynamoDB operations on hover",
   "license": "MIT",
   "author": "Justin Menga",

--- a/packages/schema/CHANGELOG.md
+++ b/packages/schema/CHANGELOG.md
@@ -1,5 +1,29 @@
 # @effect-dynamodb/schema
 
+## 1.9.2
+
+### Patch Changes
+
+- fix(schema): make the AWS-free pure-authoring path actually usable (closes #66, closes #67).
+
+  Two follow-ups to the #62 schema/runtime split, both blocking its headline use case
+  (deriving a typed aggregate input/create payload from `@effect-dynamodb/schema` with
+  no AWS SDK):
+  - **#66** — the pure edge constructors (`Aggregate.ref` / `one` / `many`) required a
+    `RefEntity` with a runtime `get` method, so aggregate edges could not be authored
+    from pure `Entity.make` definitions (which have no `get`). `RefEntity` is now the
+    minimal structural bound used only for derivation (`_tag`/`entityType`/`model`/
+    `indexes`/`schemas`); the runtime ref-hydration narrows back to a `get`-bearing
+    entity at its single call site.
+  - **#67** — `deriveAggregateSchemas` (the table-free derivation entry point) returned
+    `Schema.Top` members, so `typeof result.inputSchema.Type` collapsed to `unknown`.
+    It is now generic and returns `Schema.Codec<AggregateInputType<…>>` (plus a
+    `createSchema` alias), so the table-free path is as typed as the top-level
+    `Aggregate.make` — no stub `table` tag or GSI key config needed.
+
+  Type-checked regression tests for both land in the schema package's `tsconfig.test.json`
+  gate (now wired into `pnpm check`).
+
 ## 1.9.1
 
 ### Patch Changes

--- a/packages/schema/package.json
+++ b/packages/schema/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@effect-dynamodb/schema",
-  "version": "1.9.1",
+  "version": "1.9.2",
   "description": "Pure Effect Schema-driven entity/aggregate modeling and relationship derivation for effect-dynamodb — importable without @aws-sdk",
   "type": "module",
   "license": "MIT",
@@ -53,7 +53,7 @@
     "build": "tsc",
     "test": "vitest run",
     "test:watch": "vitest",
-    "check": "tsc --noEmit",
+    "check": "tsc --noEmit && tsc --noEmit -p tsconfig.test.json",
     "lint": "biome check .",
     "lint:fix": "biome check --write ."
   },

--- a/packages/schema/src/internal/AggregateEdges.ts
+++ b/packages/schema/src/internal/AggregateEdges.ts
@@ -22,6 +22,14 @@ import type { Schema } from "effect"
  * `indexes` mirrors `Entity.indexes` (`{ primary: { pk: { composite } }, ... }`);
  * `RefEntityIdentifierValue` reads `indexes.primary.pk.composite` to recover the
  * branded id type. A wider/looser `indexes` simply falls back to `string`.
+ *
+ * This is a PURE structural bound: edge derivation only reads
+ * `model`/`indexes`/`entityType`/`schemas` to recover the ref-id field and its
+ * (branded) identifier type — it never calls a runtime operation. It must
+ * therefore be satisfiable by a pure `EntityDefinition` (which has no `get`), so
+ * aggregate edges can be authored from `@effect-dynamodb/schema` entities with
+ * no AWS runtime (issue #66). The runtime ref-hydration path narrows back to a
+ * `get`-bearing entity at its single call site.
  */
 export interface RefEntity<M extends Schema.Top = Schema.Top, TIndexes = unknown> {
   readonly _tag: "Entity"
@@ -31,8 +39,6 @@ export interface RefEntity<M extends Schema.Top = Schema.Top, TIndexes = unknown
   readonly schemas: {
     readonly recordSchema: Schema.Codec<any>
   }
-  /** Get an entity by key — uses `any` for structural compatibility with Entity's narrower key type */
-  readonly get: (key: any) => any
 }
 
 // ---------------------------------------------------------------------------

--- a/packages/schema/src/internal/AggregateSchemas.ts
+++ b/packages/schema/src/internal/AggregateSchemas.ts
@@ -7,7 +7,7 @@
 import { Schema } from "effect"
 import { ConfiguredModelTag } from "../DynamoModel.js"
 import type { AggregateEdge, ManyEdge, RefEntity } from "./AggregateEdges.js"
-import type { BoundSubAggregate } from "./AggregateTypes.js"
+import type { AggregateInputType, BoundSubAggregate } from "./AggregateTypes.js"
 import {
   extractArrayElement as extractArrayElementImpl,
   getSchemaFields as getSchemaFieldsImpl,
@@ -28,18 +28,37 @@ import {
  * - PK composite fields are omitted (auto-generated)
  * - `Schema.toCodecJson` is applied at the end so Date fields accept ISO strings
  */
-export interface DerivedAggregateSchemas {
-  readonly inputSchema: Schema.Top
-  readonly updateSchema: Schema.Top
+/**
+ * Precisely-typed derived aggregate schemas. `TInput` is the aggregate's derived
+ * input/create payload type (see {@link AggregateInputType}); `updateSchema` is
+ * the same shape with every field optional. This lets the table-free
+ * `deriveAggregateSchemas` path be as typed as the top-level `Aggregate.make`,
+ * so an AWS-free contract package can read `typeof result.inputSchema.Type`
+ * without fabricating a stub table tag (issue #67).
+ */
+export interface DerivedAggregateSchemas<TInput = unknown> {
+  readonly inputSchema: Schema.Codec<TInput>
+  /** Alias for `inputSchema` — aggregates have no PK-composite stripping on create. */
+  readonly createSchema: Schema.Codec<TInput>
+  readonly updateSchema: Schema.Codec<Partial<TInput>>
 }
 
-export const deriveAggregateSchemas = (
-  schema: Schema.Top,
-  edges: Record<string, AggregateEdge | BoundSubAggregate<any>>,
-  pkComposites: ReadonlyArray<string>,
-): DerivedAggregateSchemas => {
+export const deriveAggregateSchemas = <
+  TSchema extends Schema.Top,
+  const TEdges extends Record<string, AggregateEdge | BoundSubAggregate<any, any>>,
+  const TPK extends ReadonlyArray<string>,
+>(
+  schema: TSchema,
+  edges: TEdges,
+  pkComposites: TPK,
+): DerivedAggregateSchemas<AggregateInputType<Schema.Schema.Type<TSchema>, TEdges, TPK>> => {
+  type Result = DerivedAggregateSchemas<
+    AggregateInputType<Schema.Schema.Type<TSchema>, TEdges, TPK>
+  >
   const fields = getSchemaFields(schema)
-  if (!fields) return { inputSchema: schema, updateSchema: schema }
+  if (!fields) {
+    return { inputSchema: schema, createSchema: schema, updateSchema: schema } as unknown as Result
+  }
 
   const omit = new Set(pkComposites)
   const newFields: Record<string, unknown> = {}
@@ -90,7 +109,10 @@ export const deriveAggregateSchemas = (
   }
   const updateSchema = Schema.toCodecJson(Schema.Struct(optionalFields as any))
 
-  return { inputSchema, updateSchema }
+  // Runtime members are `Schema.Codec<unknown>`-shaped; the precise input type
+  // is a compile-time phantom (`AggregateInputType`) — same approach the
+  // top-level `Aggregate.make` overload uses. `createSchema` aliases `inputSchema`.
+  return { inputSchema, createSchema: inputSchema, updateSchema } as unknown as Result
 }
 
 /**

--- a/packages/schema/test/use-case.test.ts
+++ b/packages/schema/test/use-case.test.ts
@@ -114,6 +114,55 @@ describe("@effect-dynamodb/schema use-case", () => {
     expect((Roster as unknown as { get?: unknown }).get).toBeUndefined()
   })
 
+  it("authors aggregate edges from pure entity definitions (#66)", () => {
+    // The pure edge constructors must accept pure `EntityDefinition`s — which
+    // have no runtime `get` — so the aggregate edge graph can be built entirely
+    // from `@effect-dynamodb/schema` (no AWS runtime). Before the fix these
+    // failed to type-check ("Property 'get' is missing"). The tsconfig.test.json
+    // gate type-checks this file, so a regression fails the build.
+    const refEdge = Aggregate.ref(Teams)
+    const oneEdge = Aggregate.one("captain", { entityType: "Player", entity: Players })
+    const manyEdge = Aggregate.many("players", { entityType: "Player", entity: Players })
+    expect(refEdge._tag).toBe("RefEdge")
+    expect(oneEdge._tag).toBe("OneEdge")
+    expect(manyEdge._tag).toBe("ManyEdge")
+
+    // ...and they compose into an aggregate authored purely from definitions.
+    const AppSchema = DynamoSchema.make({ name: "squadapp", version: 1 })
+    const Squad = Aggregate.make(Player, {
+      table: { Tag: Symbol.for("test/squad-table") },
+      schema: AppSchema,
+      pk: { field: "pk", composite: ["playerId"] },
+      collection: { index: "gsi1", name: "squad", sk: { field: "gsi1sk", composite: [] } },
+      root: { entityType: "Player" },
+      edges: { team: Aggregate.ref(Teams) },
+    })
+    expect(Squad.inputSchema).toBeDefined()
+  })
+
+  it("deriveAggregateSchemas is typed without a table tag (#67)", () => {
+    // Table-free, AWS-free: no stub `table` tag and no GSI key config — just the
+    // model, edges, and pk composites. The returned schemas must be precisely
+    // typed (NOT `Schema.Top`/`unknown`), so a contract package can read the
+    // derived create payload type directly. Also exercises #66 (pure ref edge).
+    const derived = Aggregate.deriveAggregateSchemas(Player, { team: Aggregate.ref(Teams) }, [
+      "playerId",
+    ])
+    expect(derived.inputSchema).toBeDefined()
+    expect(derived.createSchema).toBeDefined()
+    expect(derived.updateSchema).toBeDefined()
+
+    // Regression guard for #67: the derived input is a precise object type, so
+    // accessing known fields type-checks. If it collapsed to `unknown` (the bug)
+    // these would be compile errors — the schema `tsconfig.test.json` gate
+    // type-checks this file, so a regression fails the build.
+    type DerivedInput = typeof derived.inputSchema.Type
+    const readDisplayName = (i: DerivedInput): string => i.displayName
+    const readTeamId = (i: DerivedInput): string => i.teamId // ref "team" flattened to "teamId"
+    void readDisplayName
+    void readTeamId
+  })
+
   it.effect("aggregate inputSchema decodes a typed payload", () =>
     Effect.gen(function* () {
       const AppSchema = DynamoSchema.make({ name: "myapp", version: 1 })

--- a/packages/schema/tsconfig.test.json
+++ b/packages/schema/tsconfig.test.json
@@ -1,0 +1,10 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "rootDir": ".",
+    "noEmit": true,
+    "types": ["node"]
+  },
+  "include": ["src", "test/use-case.test.ts"],
+  "exclude": ["node_modules", "dist"]
+}

